### DISCUSSION TITLE
Heroku script not working

### DISCUSSION THREAD

**Discussion by mamin123** [MrMissx/Telegram_Forwarder#8] [Q&A] [answered]
I hosted my telegram forwarder using link provided on Heroku but it is not forwarding any message. Can you please help what am I missing ? I am doing this first time.

![HirokuConfig](https://user-images.githubusercontent.com/727503/108610000-1fc68e80-73a0-11eb-8615-abc485b86b60.jpg)

**Comment by mamin123** [comment]
[https://prnt.sc/102a3yr](url)
[https://prnt.sc/102a64j](url)

After further analysis in log I see below error message:

raise Exception("Your FROM_CHATS list does not contain valid integers.") using  =====>heroku logs --tail --app myappname

Chat ID used in FROM_CHATS Field is like **-1001431171975, -1001431171922**. It has minus sign in front because it is from Channel and I like to forward message coming to specific channel I am part of.

**Comment by MrMissx** [comment, marked as answer]
https://github.com/keselekpermen69/Telegram_Forwarder/blob/adc0ffd3aa8d70aae7377e69f364409d3c3a10c1/forwarder/__init__.py#L24
From here, the list of chat ids should be seperated by a `space` not a `commas`. So your FROM_CHATS vars would be like `-1001431171975 -1001431171922` etc.
The chat_id indeed should have a minus sign just leave it as it is.

**Reply by mamin123** [reply]
Thank you. Looks like error is gone. Let's see forwarding works now.

**Reply by mamin123** [reply]
Do you know how do I create copy of same script multiple times so I can use in different FROM And TO combination to forward messages ? By the way great job from one student. Keep it up you will do great.

**Reply by MrMissx** [reply]
Do you mean that you want to forward eg: from chat A to B and C to D not to all of TO_CHATS list?

In this case you don't need to multiply the script instead you should modify this https://github.com/keselekpermen69/Telegram_Forwarder/blob/adc0ffd3aa8d70aae7377e69f364409d3c3a10c1/forwarder/modules/auto_forward.py#L6-L18

to something like this
```python
if from_chat_id == <A chat_id>:
    context.bot.forward_message(chat_id=<B chat_id>, from_chat_id=from_chat_id, message_id=message.message_id)
elif from_chat_id == <C chat_id>:
   {Forward to D}
```
and so on

Hope this helps :)

**Reply by mamin123** [reply]
Still my Heroku script is not forwarding my message. Not sure what am I doing wrong. Attached is the Hiroku log. Can you please let me know what is the issue.
[mamintelegramforwarder-logs-1614125652116.txt](https://github.com/keselekpermen69/Telegram_Forwarder/files/6032485/mamintelegramforwarder-logs-1614125652116.txt)

**Reply by MrMissx** [reply]
The last line describe it clearly, *Thats Telegram thing, i can't do anything on the script

```
2021-02-24T00:13:43.678177+00:00 app[worker.1]: telegram.error.Conflict: Conflict: terminated by other getUpdates request; make sure that only one bot instance is running
```

Your bot isn't running at all. Make sure you only running **one** bot (one bot token) at a time.

**Reply by sachinbhatia4** [reply]
I am having similar issue , where i see the app is deployed fine on Heroku , i can /start and /help on the bot tooo , but forwarding is not happening , any pointers ? 

Config values seems right , i only have one from and to chat ID's with "-" sign .

**Comment by mamin123** [comment]
![Screenshot 2021-03-15 154520](https://user-images.githubusercontent.com/727503/111212374-e69bcb80-85a5-11eb-98e9-480704a200c5.jpg)
I am using Heroku and still forwarding from group to My Private group is not working. What could be wrong ?

Please let me know if you what I need to do to make it work. 

Thanks for all your help. You have been fantastic.

**Reply by mamin123** [reply]
I checked my log on heroku and I see below message:

2021-03-16T03:12:00.770782+00:00 app[worker.1]: 2021-03-16 03:12:00,770 - telegram.ext.updater - ERROR - Error while getting Updates: Conflict: terminated by other getUpdates request; make sure that only one bot instance is running
2021-03-16T03:12:01.880289+00:00 heroku[worker.1]: Process exited with status 0

Not sure why it is complaining about other GetUpdates.

**Reply by MrMissx** [reply]
I've said it clearly on my reply above. Make sure that you have **only one process running** that uses that bot token. Maybe your previous session was not propperly terminated.
[See this](https://github.com/python-telegram-bot/python-telegram-bot/issues/1879) to be more clear.

**Reply by mamin123** [reply]
I revoked my Token using botfather three times and got new one to make sure it is not being used anywhere. Could that be a issue with Heroku ? If I user it will my local Python setup, How do I keep Python Script running in background ? Is there any link you can provide me so I can follow to run script in background. I tried once with .bat file but cmd windows stays open so I want to avoid having cmd windows. Is there anyway to run as process in background ?

**Reply by MrMissx** [reply]
Not sure if its Heroku issue cause I never get that thing.
If you're on windows better to use WSL and run it on [tmux](https://github.com/tmux/tmux/wiki) it will run on background

**Reply by mamin123** [reply]
I have given up on Heroku and trying to make it work on my local PC first.

Here is what I did:
1) updated config.py file with API Key, My Chat ID, From Chat ID and To Chat ID. Everything else was kept as is.
2) Executed Command python -m forwarder on CMD. Please see output of this command. Seems working fine since no error. It also means my API and Chat ID is correct.

When I send message in one of the FROM Channel for Test it is not being forwarded to TO channel at all. The way I found channel ID is by forwarding message to @userinfobot. Whatever chat ID given here was added in config file.

Is there any way to check log or anything to see what is wrong ? Is there any test script you can provide to find out everything is configured correctly ?

<img width="590" alt="Command" src="https://user-images.githubusercontent.com/727503/111500845-ce998880-871a-11eb-85b4-81ab10a5e72c.png">
<img width="621" alt="config" src="https://user-images.githubusercontent.com/727503/111500846-ce998880-871a-11eb-8319-ea3887e11727.png">

**Reply by MrMissx** [reply]
bot need to added to both chats/channels (see #16)
